### PR TITLE
Blueprint for designing and implementing a napp message handler broker for events

### PR DESCRIPTION
**Pull request by lmarinve**
Blueprint for designing and implementing a napp message handler broker for events

Closes #issue_number

### Summary

See updated changelog file and/or add any other summarized helpful information for reviewers

### CODE REVIEW

**Review by viniarck — CHANGES_REQUESTED** [review]
@lmarinve, great to see how this is shaping up. It'll be great to have this NApp on `kytos-ng` org, and the use case involving also SDX. I get the idea of the NApp, it's understandable as it is, but this needs to be rewritten and further specified, check out the threads that I opened and the points below. Meaning, a blueprint spec, should typically be sufficient to clarify major implementation points where in future mapped tasks or PRs it's clear for whoever read the spec why certain things are the way they are, minimizing back and forth and/or potential misunderstandings of the implementation scope and/or current use case(s) in code review later on.

Here's a list of points to also be further specified:

- Should this transparently connect to a cluster too? and how secret/auth will be parametrized?
- Notice that certain events contain Python objects, which serialization are you considering that will also be compatible with consumers?
- What's the requirement for when the broker isn't available but it needs to publish a message? Will it potentially retry? Will this NApp also try to store certain events in a temporary file or will it just discard? What's acceptable for the consumers? Kytos has a basic support with a dead letter queue when an event handler happens, but once in the dead queue it expects that somebody will try to reinject it or not.
- What this NApp should do if during its `setup()` the message broker isn't available? Will this NApp require upfront that message broker is available and if it's not then just let it crash as we do for other NApps during database init? Or will it just proceed but then threat is a just temporary unavailable?
- In terms of stress tests: what's the expected number of events per second we'll consider a success? I recommend early on jump straight to this whenever possible in the protyping phase, I don't expect major surprises with `aio-pika`, and it should perform well, but let's have concrete evidence, just so it also serves as a justification why it got chosen too. 

Also, notice that the RST formatting not that well formatted, this is not a big deal, but let's make use of subsections and what not, I recommend checking on GitHub your resulting file it renders RST relatively well. One day we'll also publish blueprints on our docs, so having them structured and well displayed in sections is great.

File: docs/blueprints/EP038.rst [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+:Status: Pending
+:Type: Standards Track
+
+EP03X - Broadcasting events
+
+########
+Abstract
+########
+
+This blueprint details how Kytos should export/broadcast its events
```
[the remark this answers is not in the document]
Comment (lmarinve) [reply docs/blueprints/EP038.rst:18]: This NApp is responsible for two key functions:

Publishing Serialized KytosEvents to a Message Broker:

The primary responsibility of this NApp is to publish serialized KytosEvents to a message broker. Serialized KytosEvents contain structured data representing events within the Kytos network infrastructure.
This involves converting KytosEvents into a serialized format compatible with the message broker's communication protocol.
The NApp ensures the reliable delivery of these serialized events to the message broker, handling any necessary error handling or retry mechanisms.

Filtering KytosEvents for Publication:

Additionally, this NApp is responsible for filtering KytosEvents before publishing them to the message broker.
Filtering may involve selecting specific types of events, applying rules or criteria to determine which events are relevant for publication, or transforming events into a different format as downstream consumers require.
The NApp optimizes communication by only transmitting relevant data to the message broker by filtering events, reducing unnecessary network traffic and processing overhead.

Overall, this NApp acts as a bridge between Kytos' internal message bus and the message broker, facilitating the exchange of events between the internal network infrastructure and external systems. It guarantees the serialization, filtering, and reliable transmission of relevant events to the message broker for downstream applications or service consumption.

File: docs/blueprints/EP038.rst [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+from troubleshooting to accounting. However, creating Kytos napps to only monitor the environment should be avoided to minimize possible impacts on running the kytos instance. Moreover, this napp has the potential to expedite the prototyping of new applications and frameworks that need an understanding of the
+network topology and control plane, such as the AtlanticWave-SDX project.
+
+##########
+Requirements
+##########
+
+This napp should not be named after or delegate part of its functions to the message broker system
+This napp has to be an event broadcaster only. No input via message_bus should be supported/allowed.
+This napp has to support multiple queues: one queue per napp and a queue for all napps (verbose_queue)
```
[the remark this answers is not in the document]
Comment (lmarinve) [reply docs/blueprints/EP038.rst:36]: This NApp should support the parameterization of the exchange and routing key for publishing a given set of events. This capability allows for precise control over the distribution of events in the message broker's infrastructure. 

Exchange Configuration: An exchange is a routing mechanism that receives messages from producers and routes them to queues based on routing rules. the NApp enables the exchange specification to publish the events. 

The Napp can tailor the message routing behavior by parameterizing the exchange to suit the application's needs.

Routing Key Specification: The NApp supports parameterizing the routing keys associated with each set of events. The message broker uses a routing key as a message attribute to route messages to the appropriate queues.

The Napp defines custom routing keys based on event characteristics or metadata, allowing targeted messages to be delivered to specific queues.

Dynamic Configuration:  The NApp supports the dynamic configuration of exchange and routing key parameters, allowing the modification of these settings at runtime without requiring application restarts.

A Dynamic configuration empowers Napp to adapt message routing behavior in real-time based on changing application requirements or working conditions.

By supporting the parameterization of exchange and routing key settings, this NApp offers enhanced flexibility and is configurable in event publishing workflows. It tailors message routing behavior to align with specific use cases and optimizes message delivery and consumption within the message broker's ecosystem.

File: docs/blueprints/EP038.rst [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+Scalability and Performance Optimization:
+
+Microservices architecture facilitates resource optimization by allowing individual components to be scaled independently based on demand. This scalability ensures optimal performance even during peak loads.
+Enhanced Security:
+
+Isolating components into microservices can enhance security by reducing the attack surface area. Each microservice can have its security measures, such as access controls and authentication mechanisms, ensuring that security breaches are contained and mitigated effectively.
+
+The message queue operates by receiving messages from the events originating application and distributing them to one or more recipient applications in a first-in-first-out (FIFO) manner. This architecture facilitates communication between different parts of a system without direct coupling.
+
+Events Napp can establish separate message queues for SDX Napp and BAPM Napp to disseminate updates or commands. Events Napp would then dispatch distinct messages to each queue, and the relevant applications would retrieve messages from their designated queues.
```
[the remark this answers is not in the document]
Comment (lmarinve) [reply docs/blueprints/EP038.rst:73]: The initial concept involves the deployment of two distinct Napps: the Kytos Events Napp and the MQ Producer Messages Napp. Each of these Napps plays a unique role within the system architecture, contributing to the efficient processing and distribution of events. Let's delve into their unique functionalities:
Kytos Topology Event NApp (topology_events): Event Differentiation: This Napp differentiates various events generated within the Kytos topology. It identifies different event categories based on predefined criteria or metadata associated with each event.
After differentiating the events, the Napp serializes them into a structured format suitable for transmission and processing within the message broker ecosystem. Serialization guarantees a consistent representation of events, facilitating straightforward interpretation by downstream components.
Distribution to Requested Queues: After serialization, the Kytos Events NApp distributes the serialized events to different queues. Each queue corresponds to a specific event category or type, allowing targeted message routing and consumption by interested parties.

MQ Producer Messages NApp (mq_producer):  Replaceable RabbitMQ Code Napp: This Napp is a flexible and replaceable component responsible for interacting with RabbitMQ, a popular message broker, encapsulating the logic required for publishing messages to RabbitMQ, abstracting away the complexities of the underlying messaging infrastructure.
Message Publication: The MQ Producer Messages NApp handles publishing serialized events received from the Kytos Events NApp to RabbitMQ. It ensures reliable delivery of messages to the appropriate exchanges and queues within the RabbitMQ environment.
Configurability and Adaptability: As a replaceable component, the MQ Producer Messages NApp allows seamless integration with messaging systems or protocols beyond RabbitMQ. This design choice promotes flexibility and adaptability, enabling the system to accommodate future changes or upgrades in the messaging infrastructure.

The system can effectively manage the processing, serialization, and distribution of events generated in the Kytos network topology by employing these two Napps in conjunction. The modular architecture facilitates scalability, maintainability, and interoperability, empowering the system to evolve and meet evolving requirements and technological advancements.

File: docs/blueprints/EP038.rst [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+leveraged for multiple purposes, from capacity planning to topology evaluation,
+from troubleshooting to accounting. However, creating Kytos napps to only monitor the environment should be avoided to minimize possible impacts on running the kytos instance. Moreover, this napp has the potential to expedite the prototyping of new applications and frameworks that need an understanding of the
+network topology and control plane, such as the AtlanticWave-SDX project.
+
+##########
+Requirements
+##########
+
+This napp should not be named after or delegate part of its functions to the message broker system
+This napp has to be an event broadcaster only. No input via message_bus should be supported/allowed.
```
[the remark this answers is not in the document]
Comment (lmarinve) [reply docs/blueprints/EP038.rst:35]: The "topology_event" Napp will communicate with the topic exchange within the "mq_producer" Napp. This topic exchange operates similarly to a direct exchange but introduces a more adaptable routing mechanism based on routing patterns. Unlike direct exchanges, which rely on fixed routing keys, topic exchanges utilize wildcards for message routing, enhancing flexibility.

Here's a breakdown of its operation:

Instead of depending on specific routing keys, topic exchanges route messages by comparing a message's routing key with predefined patterns.
This comparison determines the routing of messages to one or more queues based on their correspondence with the specified patterns.
The routing key consists of a series of words separated by periods (".").
In summary, the topic exchange facilitates nuanced and dynamic message routing through wildcard patterns, offering increased flexibility and versatility in distributing messages within the messaging system.

File: docs/blueprints/EP038.rst [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+Each interdomain link has its dedicated queue.
+
+Queues are either dynamically created or configured based on settings.
+
+Producer Functionality:
+
+Upon detecting a link-down event:
+
+If the link-down event is persistent:
+The producer checks if the corresponding queue is empty.
```
[the remark this answers is not in the document]
Comment (lmarinve) [reply docs/blueprints/EP038.rst:130]: With its efficient producer-consumer relationship, the system orchestrates message handling for specific link-down events. This design ensures smooth communication, preventing message flooding and instilling confidence in its performance. Here's how it operates:

Producer Advertises Link Down: When a link-down event occurs, the producer publishes a message indicating the link's status change to down.

• Consumer's Crucial Role in Link Status Monitoring: The consumer, a key player, diligently monitors the message queue for link-down events. However, if the consumer has not yet read the message or the link status hasn't changed, the producer refrains from continuously writing messages of the same status to the queue.

Preventing Message Flooding: To prevent message flooding and conserve system resources, the producer only writes messages to the queue if the consumer has consumed the previous message or the link status has changed.

Message Continuity Based on Consumer Activity: Message continuity depends on consumer activity. If the consumer has not acknowledged or processed the previous message, the producer waits for them to read it before publishing another message with the same status. Similarly, if the link status changes, the producer updates the message accordingly.

This orchestration ensures efficient communication between the producer and consumer, preventing message overload and ensuring that messages accurately reflect the system's current state.

File: docs/blueprints/EP038.rst [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+This napp should not be named after or delegate part of its functions to the message broker system
+This napp has to be an event broadcaster only. No input via message_bus should be supported/allowed.
+This napp has to support multiple queues: one queue per napp and a queue for all napps (verbose_queue)
+This napp is optional, meaning if there is any issue with it, kytos shouldn’t be impacted
+This napp must use async.io for all asynchronous calls, for instance, alister_to instead of listen_to
+This napp has to keep consistency after reloading (re-read settings.py again, close and open rabbitmq sessions, with tests for these functionalities)
+This napp must support filtering specific REGEX messages via settings.py (a list of regex)
+
+The Broadcasting Events Napp (application) must be able to utilize any replaceable message transportation system. This requirement underscores the necessity for flexibility and adaptability in the system's architecture. Consequently, the Broadcasting Events Napp and the Events Consumer Napp should be isolated into microservices napps, distinct from the Message Queue Producer and Consumer Napps. This architectural decision offers several benefits:
```
[the remark this answers is not in the document]
Comment (lmarinve) [reply docs/blueprints/EP038.rst:42]: Separating the message broker interaction into two distinct NApps makes the system architecture more modular and easier to manage. We can develop, deploy, and scale each NApp independently, providing flexibility and adaptability.

Additionally, isolating the message broker functionality enhances fault tolerance. If one NApp experiences issues or failures, it does not directly impact the other NApp or the message broker. This isolation improves the overall robustness and reliability of the system.

Creating two NApps to isolate the message broker simplifies the system architecture, enhances modularity, and improves fault tolerance, making it easier to develop, deploy, and maintain messaging functionality.

File: docs/blueprints/EP038.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (lmarinve) [reply docs/blueprints/EP038.rst:130]: Publisher confirmations play a vital role in ensuring the reliability of message delivery. The mq_producer NApp gains confidence in successfully processing and routing a message upon receiving acknowledgment from the message broker, which prevents the mq_producer from continuously adding redundant messages to the queue. This acknowledgment mechanism ensures the reliable processing and routing of messages.
In summary, while ensuring the existence of the target exchange and setting the mandatory bit when publishing messages are essential considerations for message routing, publisher confirmation adds an extra layer of assurance for message delivery reliability.
To prevent message flooding and conserve system resources, the producer only writes messages to the queue if the consumer has consumed the previous message or the link status has changed. Message continuity is based on consumer activity; if the consumer has not acknowledged or processed the previous message, the producer waits for them to read it before publishing another message with the same status. In the same way, if the link status changes, the producer updates the message accordingly.
This orchestration ensures efficient communication between the producer and consumer, preventing message overload and ensuring that messages accurately reflect the system's current state.

File: docs/blueprints/EP038.rst [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+Exchange Configuration: An exchange is a routing mechanism that receives messages from producers and routes them to queues based on routing rules. The mq_producer NApp enables the exchange specification to publish the events.
+
+Exchange Existence Check:
+
+Before publishing messages, the mq_proucer NApp should verify whether the target exchange exists within the message broker.
+This check ensures that only legitimate exchanges receive messages, avoiding potential errors from non-existent exchanges.
+Exchange Creation if Non-existent:
+
+If the exchange does not exist, the mq_producer NApp should include logic to create the exchange dynamically.
+Even if the exchange wasn't previously defined, this dynamic creation ensures it's ready for message publishing.
```
Comment (viniarck) [inline docs/blueprints/EP038.rst:121]: This is great and can make this NApp easier to maintain. 

He prior paragraph is contradicting here though:

> Before publishing messages, the mq_proucer NApp should verify whether the target exchange exists within the message broker.

If it'll try to create, which is definitely a good pratice, there's no keep to keep checking if it exists.

File: docs/blueprints/EP038.rst [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+The NApp optimizes communication by only transmitting relevant data to the message broker by filtering events, reducing unnecessary network traffic and processing overhead.
+
+Overall, this NApp acts as a bridge between Kytos' internal message bus and the message broker, facilitating the exchange of events between the internal network infrastructure and external systems. It guarantees the serialization, filtering, and reliable transmission of relevant events to the message broker for downstream applications or service consumption.
+
+The Broadcasting Events Napp (topology_events) must be able to utilize any replaceable message transportation system. This requirement underscores the necessity for flexibility and adaptability in the system's architecture. Consequently, the Broadcasting Events Napp and the Events Consumer Napp should be isolated into microservices napps, distinct from the Message Queue Producer and Consumer Napps. This architectural decision offers several benefits:
```
Comment (viniarck) [inline docs/blueprints/EP038.rst:63]: > and the Events Consumer Napp should be isolated into microservices napps

There's no such thing as a microservice NApp. A NApp can be a "microapplication" with a defined responsiblity and you can certainly also run a kytos instance with just a single NApp acting as a sort of microservice, and for certain cases this might work when you're basically just using it to reuse `Starlette` and a NApp tiny application, but Kytos-ng is a plugin monolithic architecture. 

What you're saying about microservice generally is true and makes sense, but we NApps aren't microservices and won't be in the short or medium term, the platform itself shares many memory data structures, app instance, db instance and so on. I'm just reiterating this here just so we don't get confused or future readers don't get confused calling a NApp a microservice (by the usual software industry definition)

File: docs/blueprints/EP038.rst [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+Each interdomain link has its dedicated queue.
+Queues are either dynamically created or configured based on settings.
+
+Benefits:
+Real-time monitoring of interdomain link status.
+Scalable solution with dynamically created queues.
+Fault-tolerant design ensures persistent handling of link-down events.
+Flexibility in queue management allows dynamic addition or configuration based on settings.
+
+With its efficient producer-consumer relationship, the system orchestrates message handling for specific link-down events. This design ensures smooth communication, preventing message flooding and instilling confidence in its performance. Here's how it operates:
```
Comment (viniarck) [inline docs/blueprints/EP038.rst:172]: What is a `link-down` event? Do you mean [kytos/topology.link_down](https://github.com/kytos-ng/topology?tab=readme-ov-file#kytostopologylinkdown)? The example is great, but let's keep in mind that this be understandable for whoever is reading or will read in the future or will review PRs

File: docs/blueprints/EP038.rst [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+Producer Advertises Link Down: When a link-down event occurs, the producer publishes a message indicating the link's status change to down.
+
+Consumer's Crucial Role in Link Status Monitoring: The consumer, a key player, diligently monitors the message queue for link-down events. However, if the consumer has not yet read the message or the link status hasn't changed, the producer refrains from continuously writing messages of the same status to the queue.
+
+Preventing Message Flooding: To prevent message flooding and conserve system resources, the producer only writes messages to the queue if the consumer has consumed the previous message or the link status has changed.
+
+Message continuity is based on consumer activity; if the consumer has not acknowledged or processed the previous message, the producer waits for them to read it before publishing another message with the same status. In the same way, if the link status changes, the producer updates the message accordingly.
```
Comment (viniarck) [inline docs/blueprints/EP038.rst:180]: > the producer waits for them to read it before publishing another message with the same status. In the same way, if the link status changes, the producer updates the message accordingly.

This is coupling producer with a consumer. It sounds strange to me. I've never seen any type of pub sub where a publisher were to do it. It's also contradicting some of the prior information about producer being independent from a consumer (as it should be). Can you double check this paragraph?

File: docs/blueprints/EP038.rst [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+
+Preventing Message Flooding: To prevent message flooding and conserve system resources, the producer only writes messages to the queue if the consumer has consumed the previous message or the link status has changed.
+
+Message continuity is based on consumer activity; if the consumer has not acknowledged or processed the previous message, the producer waits for them to read it before publishing another message with the same status. In the same way, if the link status changes, the producer updates the message accordingly. 
+
+Publisher confirmations play a vital role in ensuring the reliability of message delivery. The mq_producer NApp gains confidence in successfully processing and routing a message upon receiving acknowledgment from the message broker, which prevents the mq_producer from continuously adding redundant messages to the queue. This acknowledgment mechanism ensures the reliable processing and routing of messages.
```
Comment (viniarck) [inline docs/blueprints/EP038.rst:182]: > routing a message upon receiving acknowledgment from the message broker, which prevents the mq_producer from continuously adding redundant messages to the queue.

This is good, but let's specify which protocol attrs we're talking here. In a PR this will need to be done and then you can refer to it. As of now this is just high level speaking about it, I have no idea if you're meaning to leverage the existing protocol flags to do it or roll out an implementation.

**Review by viniarck — COMMENTED** [review]
@lmarinve, good progress. Some of updates clarified certain questions, but there are major requirements and implementation requirements that need to be sorted out. Notice that my main concerns and questions aren't why we're building this NApp (that's already clear in the doc), but how certaing things will be built and how it'll be implemented, which should serve as a basis for future tasks that will be broken down and future reviewed in PRs facilitating for you and for whoever will also review, just so there's no misexpectactions in PRs why certain thing is being implemented in a certain way. That's why I'm being extra pedantic here, If I read a paragraph or requirement and can't understand how it'll be implementend in this iteration, I'll keep asking for changes to be better specified. Happy to also discuss in a call if there's still misunderstanding on why I'm asking these things here, ping me if needed.

Finally, a trivial detail, notice that this `.rst` doc here is barely using any RST formatting, it's not displaying nicely, you can have a [rendered view here](https://github.com/kytos-ng/kytos/blob/457-blueprint-for-a-message-broker-bridge-napp/docs/blueprints/EP038.rst), not a big deal, but I'll recommend that you try to format with subsections, bullet points and so on.

File: docs/blueprints/EP038.rst [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+
+##########
+Requirements
+##########
+
+This napp should not be named after or delegate part of its functions to the message broker system
+This napp has to be an event broadcaster only. No input via message_bus should be supported/allowed.
+This napp has to support multiple queues: one queue per napp and a queue for all napps (verbose_queue)
+This napp is optional, meaning if there is any issue with it, kytos shouldn’t be impacted
+This napp must use async.io for all asynchronous calls, for instance, alister_to instead of listen_to
```
[the remark this answers is not in the document]
Comment (lmarinve) [reply docs/blueprints/EP038.rst]: I added the requirement in the last commit.

File: docs/blueprints/EP038.rst [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+##########
+Requirements
+##########
+
+This napp has to be an event broadcaster only. No input via message_bus should be supported/allowed.
+This napp has to support multiple queues: one queue per napp and a queue for all napps (verbose_queue)
+This napp is optional, meaning if there is any issue with it, kytos shouldn’t be impacted
+This napp must use async.io for all asynchronous calls, for instance, alister_to instead of listen_to
+This napp has to keep consistency after reloading (re-read settings.py again, close and open rabbitmq sessions, with tests for these functionalities)
+This napp must support filtering specific REGEX messages via settings.py (a list of regex)
```
[the remark this answers is not in the document]
Comment (lmarinve) [reply docs/blueprints/EP038.rst]: Setting file has the following events

ADMIN_EVENTS = [
"version/control.initialize",
"kytos/topology.switch.enabled",
"kytos/topology.switch.disabled",
"kytos/topology.switch.metadata.added",
"kytos/topology.interface.metadata.added",
"kytos/topology.link.metadata.added",
"kytos/topology.switch.metadata.removed",
"kytos/topology.interface.metadata.removed",
"kytos/topology.link.metadata.removed",
]

OPERATIONAL_EVENTS = [
"topology_loaded",
"kytos/topology.link_up",
"kytos/topology.link_down",
]

File: docs/blueprints/EP038.rst [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+
+Message queues decouple components within the system. Events Napp can transmit updates without the continuous availability of SDX Napp or BAPM applications. Moreover, the persistent nature of message queues ensures that if any application experiences a restart, it can seamlessly resume processing messages from its designated queue once it is back online.
+
+This approach enhances scalability and fault tolerance in applications by reducing dependencies between interconnected systems. Additionally, it facilitates better handling of system failures or temporary unavailability, thereby strengthening the overall robustness of the architecture.
+
+AMQP 0.9.1, a highly efficient and versatile protocol, empowers RabbitMQ, a widely acclaimed message broker, to communicate seamlessly across various systems. This robust combination ensures reliable communication between different components of a distributed system.
+
+aio-pika is an asynchronous AMQP client library designed for Python applications. It enables asynchronous and efficient interaction with RabbitMQ, making it well-suited for high-performance applications or systems that require non-blocking I/O operations.
```
[the remark this answers is not in the document]
Comment (lmarinve) [reply docs/blueprints/EP038.rst]: https://aio-pika.readthedocs.io/en/latest/rabbitmq-tutorial/1-introduction.html

File: docs/blueprints/EP038.rst [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+With Kytos running in production, we learned that Kytos events can be
+leveraged for multiple purposes, from capacity planning to topology evaluation,
+from troubleshooting to accounting. However, creating Kytos napps to only monitor the environment should be avoided to minimize possible impacts on running the kytos instance. Moreover, this napp has the potential to expedite the prototyping of new applications and frameworks that need an understanding of the
+network topology and control plane, such as the AtlanticWave-SDX project.
+
+##########
+Requirements
+##########
+
+This napp should not be named after or delegate part of its functions to the message broker system
```
[the remark this answers is not in the document]
Comment (lmarinve) [reply docs/blueprints/EP038.rst]: The item in question has been removed from the blueprint document.

File: docs/blueprints/EP038.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (lmarinve) [reply docs/blueprints/EP038.rst]: 2.- mq_producer NApp
The mq_producer NApp should support parameterizing the exchange and routing key for publishing a given set of events. This capability allows for precise control over the distribution of events in the message broker's infrastructure.

This topic exchange operates similarly to a direct exchange but introduces a more adaptable routing mechanism based on routing patterns. Unlike direct exchanges, which rely on fixed routing keys, topic exchanges utilize wildcards for message routing, enhancing flexibility.

Here's a breakdown of its operation:

Instead of depending on specific routing keys, topic exchanges route messages by comparing a message's routing key with predefined patterns.

This comparison determines the routing of messages to one or more queues based on their correspondence with the specified patterns.

The routing key consists of a series of words separated by periods (".").

In summary, the topic exchange facilitates nuanced and dynamic message routing through wildcard patterns, offering increased flexibility and versatility in distributing messages within the messaging system.

File: docs/blueprints/EP038.rst [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+Benefits
+#######
+
+Real-time monitoring of interdomain link status.
+
+Scalable solution with dynamically created queues.
+
+Fault-tolerant design ensures persistent handling of link-down events.
+
+Flexibility in queue management, allowing for dynamic addition or configuration based on settings.
```
[the remark this answers is not in the document]
Comment (lmarinve) [reply docs/blueprints/EP038.rst]: Exchange Configuration
An exchange is a routing mechanism that receives messages from producers and routes them to queues based on routing rules. The mq_producer NApp enables the exchange specification to publish the events.

Exchange Existence Check
At setup() method of mq_producer NApp should verify whether the target exchange exists within the message broker. This check ensures that only legitimate exchanges receive messages, avoiding potential errors from non-existent exchanges.

Exchange Creation if Non-existent
If the exchange does not exist, the mq_producer NApp should include logic to create the exchange dynamically. Even if the exchange wasn't previously defined, this dynamic creation ensures it's ready for message publishing.

Error Handling for Exchange Creation
We should implement proper error handling to address scenarios where exchange creation fails. The NApp should handle exceptions gracefully and potentially log relevant error messages for troubleshooting purposes. By incorporating these specifications, the mq_producer NApp can manage the target exchange effectively, ensuring reliable message publishing and providing resilience in dynamic exchange creation.

The mq_producer Napp can tailor the message routing behavior by parameterizing the exchange to suit the application's needs.

File: docs/blueprints/EP038.rst [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+
+Upon detecting a link-down event:
+
+If the link-down event is persistent:
+The producer checks if the corresponding queue is empty.
+
+If empty, it adds a new message "[Link Down]" to the unique element link status queue.
+
+If the link-down event is not persistent:
+The producer keeps checking until the consumer reads and deletes the message.
```
[the remark this answers is not in the document]
Comment (lmarinve) [reply docs/blueprints/EP038.rst]: Consumer's Crucial Role in Link Status Monitoring:

The consumer, a key player, diligently monitors the message queue for link-down events. However, if the consumer still needs to read the message or the link status hasn't changed, the producer refrains from continuously writing messages of the same status to the queue.

Preventing Message Flooding:

To prevent message flooding and conserve system resources, the producer only writes messages to the queue if the consumer has consumed the previous message or the link status has changed.

Message continuity is based on consumer activity:

If the consumer has yet to acknowledge or process the previous message, the producer waits for them to read it before publishing another message with the same status. In the same way, if the link status changes, the producer updates the message accordingly.

Publisher confirmations play a vital role in ensuring the reliability of message delivery:

The mq_producer NApp gains confidence in successfully processing and routing a message upon receiving acknowledgment from the message broker, which prevents the mq_producer from continuously adding redundant messages to the queue. This acknowledgment mechanism ensures the reliable processing and routing of messages.

In summary, while ensuring the existence of the target exchange and setting the mandatory bit when publishing messages are essential considerations for message routing, publisher confirmation adds an extra layer of assurance for message delivery reliability.

Overall, this use case demonstrates how message queues can be effectively utilized for monitoring and managing interdomain link status changes in an SDX network infrastructure, ensuring timely detection and response to connectivity issues.

File: docs/blueprints/EP038.rst [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+
+Consumer Functionality:
+
+Consumers must be aware of the queues they need to monitor for link status changes.
+
+Consumers continuously monitor the designated queues for new messages indicating link status changes.
+
+Upon receiving a message:
+
+If the message is "[Link Down]," the consumer processes the link down event.
```
[the remark this answers is not in the document]
Comment (lmarinve) [reply docs/blueprints/EP038.rst]: OPERATIONAL_EVENTS = [
"topology_loaded",

"kytos/topology.link_up",

"kytos/topology.link_down",

]

File: docs/blueprints/EP038.rst [hunk 19]
```diff
@@ -0,0 +1,460 @@
...
+The Kytos-sdx-topology Napp should actively monitor topology changes by subscribing to events using the Kytos Event decorated function. This design choice must ensure the system stays updated with any modifications in the Kytos-ng topology and enables real-time processing of topology-related events.
+
+Kytos-sdx-topology kytos Napp
+https://github.com/atlanticwave-sdx/kytos-sdx-topology
+
+Actual kytos-sdx-topology napp is designed to do the whole process in one napp, In atlanticwave sdx infrastructure, it's crucial to monitor the status of interdomain links for connectivity and reliability. This use case involves setting up a system to detect link status changes (up/down) and notify consumers about these events. 
+
+The purpose of this blueprint is to design and evaluate the introduction of a message queue and separate the process as follows:
+
+Creating two NApps to separate the event listener and treatment from the post office message broker simplifies the system architecture, enhances modularity, and improves fault tolerance. Here's how:
```
Comment (viniarck) [inline docs/blueprints/EP038.rst:52]: Luis, this part is a bit confusing. Maybe you're already picturing that you'll have a consumer NApp on a different kytosd instance. The consumer on the other end could be any kind of consumer whether a NApp on another instance, or BAPM, or whatever else application, or even another application in a different language (and that's why I'm also asking to specify the serialization that will be used, whether or not the consumer will be needed to be in Python or not):



For whoever is reading this doc, the "two NApps" here are kind of not making it clear, and as if this NApp here were to pretty much act as a parallel event bus, which it can be, but the thing is, if one day we decide to also support a richer message broker to send the events it'll be baked in the core and not in a NApp. So, for this NApp here, as you're documenting, it's a adapt to bridge events from the internal bus to an external (RabbitMQ) message broker. 

That said, this spec should focus on the producer, of course you can also briefly hightlight how a consumer would look like it you want to use a consumer in the SDX, but users shouldn't be constrained to always use "two NApps" to have events being bridged to RabbitMQ, just a single NApp that you're specifying should suffice. Now, if you with to make official the consumer in another blueprint related to SDX, it's up to you. That's probably already clear to you, but just highlighting this point here.

**Review by viniarck — CHANGES_REQUESTED** [review]
Luis, I realize that you're pushing commits and still changing the text, but I took the opportunity to do another partial code review. I'll do a final one when you need. But, I figured this partial one can still make sure we're on the same page regarding the points that I'm still asking to be specified.

**Comment by viniarck** [comment]
I'll go ahead and close this PR. 

Luis, when you're ready and addressed the comments, send a new one, it'll be easier that way. Thanks.
